pkg/reconciler/broker/targets.go

```diff
@@ -0,0 +1,205 @@
+package broker
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/google/knative-gcp/pkg/broker/config"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/typed/core/v1"
+	"knative.dev/eventing/pkg/logging"
+)
+
+const (
+	targetsCMKey          = "targets"
+	targetsFileName       = "targets.txt"
+	targetsCMResyncPeriod = 10 * time.Second
+)
+
+type BrokerConfigUpdater interface {
+	UpdateBrokerConfig(context.Context, *config.Broker) error
+}
+
+type updateReq struct {
+	broker  *config.Broker
+	replyCh chan error
+}
+
+type targetsReconciler struct {
+	targetsName, targetsNS string
+	targets                *config.TargetsConfig
+	updateCh               chan updateReq
+	configMaps             v1.ConfigMapInterface
+	logger                 *zap.Logger
+}
+
+func NewTargetsReconciler(ctx context.Context, client kubernetes.Interface, targetsNS, targetsName string) BrokerConfigUpdater {
+	r := &targetsReconciler{
+		targetsName: targetsName,
+		targetsNS:   targetsNS,
+		updateCh:    make(chan updateReq),
+		configMaps:  client.CoreV1().ConfigMaps(targetsNS),
+		logger:      logging.FromContext(ctx),
+	}
+	if err := r.loadTargetsConfig(ctx); err != nil {
+		r.logger.Error("error loading targets config", zap.Error(err))
+		r.targets = new(config.TargetsConfig)
+	}
+	go r.reconcile(ctx)
+	return r
+}
+
+func (r *targetsReconciler) UpdateBrokerConfig(ctx context.Context, b *config.Broker) error {
```

Comment:
I _think_ the complexity here is aimed at allowing multiple threads to call this method concurrently **and** get an error back (which the previous implementation didn't allow).

Reply:
Yes, this allows the serialization and batching of writes across multiple reconcile goroutines while providing a blocking interface to the caller.